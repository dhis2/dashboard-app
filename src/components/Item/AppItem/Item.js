import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import NotInterestedIcon from '@material-ui/icons/NotInterested';

import { FILTER_USER_ORG_UNIT } from '../../../actions/itemFilter';
import { sGetItemFilterRoot } from '../../../reducers/itemFilter';
import ItemHeader from '../ItemHeader';
import Line from '../../../widgets/Line';

const getIframeSrc = (appDetails, item, itemFilter) => {
    let iframeSrc = `${appDetails.launchUrl}?dashboardItemId=${item.id}`;

    if (
        itemFilter &&
        itemFilter[FILTER_USER_ORG_UNIT] &&
        itemFilter[FILTER_USER_ORG_UNIT].length
    ) {
        const ouIds = itemFilter[FILTER_USER_ORG_UNIT].map(
            ouPath => ouPath.split('/').slice(-1)[0]
        );

        iframeSrc += `&userOrgUnit=${ouIds.join(',')}`;
    }

    return iframeSrc;
};

const AppItem = ({ item, itemFilter }, context) => {
    let appDetails;

    const appKey = item.appKey;

    if (appKey) {
        appDetails = context.d2.system.installedApps.find(
            app => app.key === appKey
        );
    }

    return appDetails && appDetails.name && appDetails.launchUrl ? (
        <Fragment>
            <ItemHeader title={appDetails.name} />
            <Line />
            <iframe
                title={appDetails.name}
                src={getIframeSrc(appDetails, item, itemFilter)}
                className="dashboard-item-content"
                style={{ border: 'none' }}
            />
        </Fragment>
    ) : (
        <Fragment>
            <ItemHeader title={`${appKey} app not found`} />
            <Line />
            <div
                className="dashboard-item-content"
                style={{
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    height: '90%',
                }}
            >
                <NotInterestedIcon
                    color="disabled"
                    disabled
                    style={{ width: 100, height: 100, align: 'center' }}
                />
            </div>
        </Fragment>
    );
};

AppItem.contextTypes = {
    d2: PropTypes.object,
};

const mapStateToProps = state => ({
    itemFilter: sGetItemFilterRoot(state),
});

export default connect(mapStateToProps)(AppItem);
