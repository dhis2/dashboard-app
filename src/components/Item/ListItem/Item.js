import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { colors } from '@dhis2/ui-core'
import DescriptionIcon from '../../../icons/Description'

import DeleteIcon from '../../../icons/Delete'
import Line from '../../../widgets/Line'
import { itemTypeMap, getItemUrl } from '../../../modules/itemTypes'
import { orArray } from '../../../modules/util'
import { tRemoveListItemContent } from './actions'
import ItemHeader from '../ItemHeader'
import classes from './Item.module.css'

const getItemTitle = item => itemTypeMap[item.type].pluralTitle

const getContentItems = item =>
    orArray(item[itemTypeMap[item.type].propName]).filter(
        (item, index, array) =>
            array.findIndex(el => el.id === item.id) === index
    )

const ListItem = (props, context) => {
    const { item, editMode, tRemoveListItemContent } = props
    const contentItems = getContentItems(item)

    const getLink = contentItem => {
        const deleteButton = (
            <button
                className={classes.deletebutton}
                onClick={() => tRemoveListItemContent(item, contentItem)}
            >
                <DeleteIcon className={classes.deleteicon} />
            </button>
        )

        return (
            <>
                <a
                    className={classes.link}
                    style={{ color: colors.grey900 }}
                    href={getItemUrl(item.type, contentItem, context.d2)}
                >
                    {contentItem.name}
                </a>
                {editMode ? deleteButton : null}
            </>
        )
    }

    return (
        <>
            <ItemHeader title={getItemTitle(item)} itemId={item.id} />
            <Line />
            <div className="dashboard-item-content">
                <ul className={classes.list}>
                    {contentItems.map(contentItem => (
                        <li className={classes.item} key={contentItem.id}>
                            <DescriptionIcon className={classes.itemicon} />
                            {getLink(contentItem)}
                        </li>
                    ))}
                </ul>
            </div>
        </>
    )
}

ListItem.propTypes = {
    editMode: PropTypes.bool,
    item: PropTypes.object,
    tRemoveListItemContent: PropTypes.func,
}

ListItem.contextTypes = {
    d2: PropTypes.object,
}

export default connect(null, {
    tRemoveListItemContent,
})(ListItem)
