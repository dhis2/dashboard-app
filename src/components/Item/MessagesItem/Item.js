import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import { useConfig } from '@dhis2/app-runtime'
import ItemHeader from '../ItemHeader/ItemHeader'
import Line from '../../../widgets/Line'
import { useUserSettings } from '../../UserSettingsProvider'

import { sGetMessagesRoot } from '../../../reducers/messages'
import { formatDate } from '../../../modules/util'
import { isViewMode } from '../../Dashboard/dashboardModes'

import classes from './styles/Item.module.css'
import './MessagesItem.css'

const PRIVATE = 'PRIVATE'

const messageTypes = {
    [PRIVATE]: 'Private',
    VALIDATION_RESULT: 'Validation',
    TICKET: 'Ticket',
    SYSTEM: 'System',
}

const MessagesItem = ({ messages, item, dashboardMode }) => {
    const { baseUrl } = useConfig()
    const { userSettings } = useUserSettings()

    const getMessageHref = msg => {
        const msgIdentifier = msg ? `#/${msg.messageType}/${msg.id}` : ''
        return `${baseUrl}/dhis-web-messaging/${msgIdentifier}`
    }

    const getMessageSender = msg => {
        const latestMsg = msg.messages.slice(-1)[0]
        return latestMsg.sender ? latestMsg.sender.displayName : ''
    }

    const getMessageItems = () => {
        const modeClass = isViewMode(dashboardMode) ? 'view' : null

        return messages.map(msg => {
            const redirectToMsg = () => {
                if (isViewMode(dashboardMode)) {
                    document.location.href = getMessageHref(msg)
                }
            }

            const sender =
                msg.messageType === PRIVATE
                    ? getMessageSender(msg)
                    : messageTypes[msg.messageType]

            const readClass = !msg.read ? 'unread' : null
            const latestMsg = msg.messages.slice(-1)[0]
            const msgDate = latestMsg.lastUpdated

            return (
                <li
                    className={`message-item ${modeClass}`}
                    key={msg.id}
                    onClick={redirectToMsg}
                >
                    <p className={`message-title ${readClass}`}>
                        {msg.displayName} ({msg.messageCount})
                    </p>
                    <p className={classes.sender}>
                        {sender} -{' '}
                        {formatDate(msgDate, userSettings.keyUiLocale)}
                    </p>
                    <p className={classes.snippet}>{latestMsg.text}</p>
                </li>
            )
        })
    }

    return (
        <>
            <ItemHeader
                title={i18n.t('Messages')}
                itemId={item.id}
                dashboardMode={dashboardMode}
                isShortened={item.shortened}
            />
            <Line />
            {messages.length > 0 && (
                <div className="dashboard-item-content">
                    <ul className={classes.list}>{getMessageItems()}</ul>
                    <div className={classes.seeAll}>
                        <a href={getMessageHref()}>
                            {i18n.t('See all messages')}
                        </a>
                    </div>
                </div>
            )}
        </>
    )
}

MessagesItem.propTypes = {
    dashboardMode: PropTypes.string,
    item: PropTypes.object,
    messages: PropTypes.array,
}

const mapStateToProps = state => {
    return {
        messages: Object.values(sGetMessagesRoot(state)),
    }
}

const MessagesContainer = connect(mapStateToProps, null)(MessagesItem)

export default MessagesContainer
