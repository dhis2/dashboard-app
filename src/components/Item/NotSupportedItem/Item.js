import React, { Fragment } from 'react';
import ItemHeader from '../ItemHeader';
import SvgIcon from 'd2-ui/lib/svg-icon/SvgIcon';

const NotSupportedItem = props => (
    <Fragment>
        <ItemHeader title={`Item type not supported: ${props.item.type}`} />
        <div
            style={{
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                height: '90%',
            }}
        >
            <SvgIcon
                icon="NotInterested"
                style={{ width: 100, height: 100, align: 'center' }}
                disabled
            />
        </div>
    </Fragment>
);

export default NotSupportedItem;
