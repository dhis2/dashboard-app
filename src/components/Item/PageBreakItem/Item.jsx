import React from 'react'

const PageBreakItem = () => <div />

export default PageBreakItem
