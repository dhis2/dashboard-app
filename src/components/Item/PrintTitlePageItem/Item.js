import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'

import {
    sGetSelectedId,
    sGetSelectedShowDescription,
} from '../../../reducers/selected'
import {
    sGetDashboardById,
    EMPTY_DASHBOARD,
} from '../../../reducers/dashboards'
import { sGetNamedItemFilters } from '../../../reducers/itemFilters'

import classes from './styles/Item.module.css'

const PrintTitlePageItem = ({
    name,
    description,
    itemFilters,
    showDescription,
}) => {
    const getItemFilterList = () => {
        const listItems = itemFilters.map(({ name, values }) => (
            <li className={classes.filterListItem} key={name}>
                {name}: {values.map(val => val.name).join(', ')}
            </li>
        ))

        return <ul className={classes.filterList}>{listItems}</ul>
    }

    return (
        <div className={classes.titlePage}>
            <p className={classes.name}>{name}</p>
            {showDescription && description && (
                <p className={classes.description}>{description}</p>
            )}
            {itemFilters.length > 0 && (
                <>
                    <p className={classes.filterTitle}>
                        {i18n.t('Filters applied')}
                    </p>
                    {getItemFilterList()}
                </>
            )}
        </div>
    )
}

PrintTitlePageItem.propTypes = {
    description: PropTypes.string,
    itemFilters: PropTypes.array,
    name: PropTypes.string,
    showDescription: PropTypes.bool,
}

PrintTitlePageItem.defaultProps = {
    description: '',
    name: '',
    showDescription: false,
}

const mapStateToProps = state => {
    const id = sGetSelectedId(state)
    const dashboard = sGetDashboardById(state, id) || EMPTY_DASHBOARD

    return {
        name: dashboard.displayName,
        itemFilters: sGetNamedItemFilters(state),
        description: dashboard.displayDescription,
        showDescription: sGetSelectedShowDescription(state),
    }
}

export default connect(mapStateToProps)(PrintTitlePageItem)
