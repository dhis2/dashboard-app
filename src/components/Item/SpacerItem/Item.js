import React, { Fragment } from 'react';
import i18n from '@dhis2/d2-i18n';

import { colors } from '@dhis2/ui-core';
import ItemHeader from '../ItemHeader';

const style = {
    margin: '21px 28px',
    fontSize: '14px',
    lineHeight: '18px',
    color: colors.grey600,
};

const SpacerItem = () => {
    return (
        <Fragment>
            <ItemHeader title={i18n.t('Spacer')} />
            <p style={style}>
                {i18n.t(
                    'Use a spacer to create empty vertical space between other dashboard items.'
                )}
            </p>
        </Fragment>
    );
};

export default SpacerItem;
