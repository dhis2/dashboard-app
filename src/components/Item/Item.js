import React from 'react';

import VisualizationItem from './VisualizationItem/Item';
import MessagesItem from './MessagesItem/Item';
import NotSupportedItem from './NotSupportedItem/Item';
import ListItem from './ListItem/Item';
import TextItem from './TextItem/Item';
import AppItem from './AppItem/Item';
import SpacerItem from './SpacerItem/Item';
import {
    APP,
    REPORT_TABLE,
    CHART,
    MAP,
    EVENT_CHART,
    EVENT_REPORT,
    MESSAGES,
    REPORTS,
    RESOURCES,
    USERS,
    TEXT,
    SPACER,
} from '../../modules/itemTypes';
import { DEFAULT_STATE_ITEM_FILTERS } from '../../reducers/itemFilters';

const getGridItem = type => {
    switch (type) {
        case REPORT_TABLE:
        case CHART:
        case MAP:
        case EVENT_CHART:
        case EVENT_REPORT:
            return VisualizationItem;
        case MESSAGES:
            return MessagesItem;
        case REPORTS:
        case RESOURCES:
        case USERS:
            return ListItem;
        case TEXT:
            return TextItem;
        case SPACER:
            return SpacerItem;
        case APP:
            return AppItem;
        default:
            return NotSupportedItem;
    }
};

export const Item = props => {
    const GridItem = getGridItem(props.item.type);

    return (
        <GridItem
            item={props.item}
            editMode={props.editMode}
            itemFilters={
                props.editMode ? DEFAULT_STATE_ITEM_FILTERS : props.itemFilters
            }
            onToggleItemExpanded={props.onToggleItemExpanded}
        />
    );
};
