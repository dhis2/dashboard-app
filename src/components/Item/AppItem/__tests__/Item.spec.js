import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import WindowDimensionsProvider from '../../../WindowDimensionsProvider.js'
import Item from '../../Item.js'

jest.mock('@dhis2/analytics', () => ({
    useCachedDataQuery: () => ({
        apps: [
            {
                key: 'scorecard',
                name: 'Scorecard',
                launchUrl: 'launchurl',
                pluginLaunchUrl: 'pluginLaunchUrl',
            },
            {
                key: 'noTitle',
                name: 'No Title',
                launchUrl: 'launchurl',
                pluginLaunchUrl: 'pluginLaunchUrl',
                settings: {
                    dashboardWidget: {
                        hideTitle: true,
                    },
                },
            },
        ],
        currentUser: {
            username: 'rainbowDash',
            id: 'r3nb0d5h',
        },
    }),
    getDimensionById: jest.fn(),
}))

jest.mock(
    '../../ItemHeader/DeleteItemButton.js',
    () =>
        function Mock() {
            return <div className="DeleteItemButton" />
        }
)

const mockStore = configureMockStore()

const item = {
    appKey: 'scorecard',
    id: 'rainbowdash',
    shortened: false,
}

const itemWithoutTitle = {
    appKey: 'noTitle',
    id: 'twilightsparkle',
    shortened: false,
}

test('renders a valid App item in view mode', () => {
    const store = {
        itemFilters: {},
    }
    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={item} dashboardMode={'view'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a valid App item with filter in view mode', () => {
    const store = {
        itemFilters: {
            ou: [{ path: '/rainbow' }],
        },
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={item} dashboardMode={'view'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a valid App item with filter in edit mode', () => {
    const store = {
        itemFilters: {
            ou: [{ path: '/rainbow' }],
        },
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={item} dashboardMode={'edit'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a valid App item without title in view mode if specified in app settings', () => {
    const store = {
        itemFilters: {},
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={itemWithoutTitle} dashboardMode={'view'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders a valid App item with title in edit mode irrespective of app settings', () => {
    const store = {
        itemFilters: {},
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={itemWithoutTitle} dashboardMode={'edit'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('renders an invalid App item', () => {
    const store = {
        itemFilters: {
            ou: [{ path: '/rainbow' }],
        },
    }

    const invalidItem = {
        appKey: 'unknownApp',
        id: 'unknown',
        shortened: false,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <WindowDimensionsProvider>
                <Item item={invalidItem} dashboardMode={'edit'} />
            </WindowDimensionsProvider>
        </Provider>
    )
    expect(container).toMatchSnapshot()
})
