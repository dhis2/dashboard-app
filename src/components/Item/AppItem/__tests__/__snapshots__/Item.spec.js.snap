// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a valid App item in view mode 1`] = `
<div>
  <div
    class="itemHeaderWrap"
  >
    <p
      class="itemTitle"
    >
      Scorecard
    </p>
  </div>
  <hr
    style="background-color: rgb(243, 245, 247); height: 1px; margin: 0px 0px 5px 0px;"
  />
  <iframe
    class="dashboard-item-content"
    src="launchurl?dashboardItemId=rainbowdash"
    title="Scorecard"
  />
</div>
`;

exports[`renders a valid App item with filter in edit mode 1`] = `
<div>
  <div
    class="itemHeaderWrap"
  >
    <p
      class="itemTitle"
    >
      Scorecard
    </p>
    <div
      class="itemActionsWrap"
    >
      <button
        class="deleteItemButton"
        title="Delete item"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          style="fill: #d32f2f;"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
  <hr
    style="background-color: rgb(243, 245, 247); height: 1px; margin: 0px 0px 5px 0px;"
  />
  <iframe
    class="dashboard-item-content"
    src="launchurl?dashboardItemId=rainbowdash"
    title="Scorecard"
  />
</div>
`;

exports[`renders a valid App item with filter in view mode 1`] = `
<div>
  <div
    class="itemHeaderWrap"
  >
    <p
      class="itemTitle"
    >
      Scorecard
    </p>
  </div>
  <hr
    style="background-color: rgb(243, 245, 247); height: 1px; margin: 0px 0px 5px 0px;"
  />
  <iframe
    class="dashboard-item-content"
    src="launchurl?dashboardItemId=rainbowdash&userOrgUnit=rainbow"
    title="Scorecard"
  />
</div>
`;

exports[`renders an invalid App item 1`] = `
<div>
  <div
    class="itemHeaderWrap"
  >
    <p
      class="itemTitle"
    >
      unknownApp app not found
    </p>
  </div>
  <hr
    style="background-color: rgb(243, 245, 247); height: 1px; margin: 0px 0px 5px 0px;"
  />
  <div
    class="dashboard-item-content"
    style="display: flex; justify-content: center; align-items: center; height: 90%;"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-10 MuiSvgIcon-colorDisabled-15"
      disabled=""
      focusable="false"
      role="presentation"
      style="width: 100px; height: 100px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8 0-1.85.63-3.55 1.69-4.9L16.9 18.31C15.55 19.37 13.85 20 12 20zm6.31-3.1L7.1 5.69C8.45 4.63 10.15 4 12 4c4.42 0 8 3.58 8 8 0 1.85-.63 3.55-1.69 4.9z"
      />
    </svg>
  </div>
</div>
`;
